File: core/src/main/java/org/lei/opi/core/Jovp.java
package org.lei.opi.core;

import java.io.IOException;
import java.util.Arrays;
import java.util.ArrayList;
import java.util.HashMap;

import org.lei.opi.core.OpiListener.Command;
import org.lei.opi.core.definitions.Packet;
import org.lei.opi.core.definitions.Parameter;

import es.optocom.jovp.definitions.ViewMode;

/**
 * JOVP client - will send messages to JOVP server...
 *
 * @since 0.0.1
 */
public abstract class Jovp extends OpiMachine {

    public static class Settings extends OpiMachine.Settings {
        public int screen;
        public int[] physicalSize;
        public boolean pseudoGray;
        public boolean fullScreen;
        public int distance;
        public String viewMode;
        public String input;
        public boolean tracking;
        public String gammaFile;

        public void setScreen(int screen) { this.screen = screen; }
        public void setPhysicalSize(int[] psize) { this.physicalSize = psize; }
        public void setViewMode(String s) { this.viewMode = s; }
    };

    /** Settings */
    protected Settings settings;
    public Settings getSettings() { return this.settings; }

    /**
     * NOTE: Does not create connection to actual machine via settings-> ip:port
     * @param parentScene The Scene to return to when this object is closed.
     * @throws InstantiationException if cannot connect to JOVP server
     */
    public Jovp(javafx.scene.Scene parentScene) throws InstantiationException {
        super(parentScene);
        this.settings = (Settings) OpiMachine.fillSettings(this.getClass().getSimpleName());
       
        setVFCanvas(
          settings.viewMode.toLowerCase().equals(ViewMode.MONO.toString().toLowerCase()),
          settings.tracking
        );
    }

    // create instance AND open connection to settings-> ip port
    public Jovp(javafx.scene.Scene parentScene, boolean createConnection) throws InstantiationException {
        this(parentScene);
        if (!this.connect(settings.ip, settings.port))
            throw new InstantiationException(String.format("Cannot connect to %s:%s", settings.ip, settings.port));
    }

    /**
    * opiInitialise: send initialization code to JOVP on the machine and get results
    * 
    * @param args A map of name:value pairs for Params these are used to override any of the default initConfiguration().
    * 
    * @return A Packet that is returned from the machine. 
    * SIDE EFFECT: with return JSON object with machine specific initialise information
    * 
    * @since 0.0.1
    */
    public Packet initialize(HashMap<String, Object> args) {
        try {
            this.send(initConfiguration());
            Packet p = this.receive();
            return Packet.checkReturnElements(p, this.opiMethods, "initialize");
        } catch (IOException e) {
            return Packet.error(COULD_NOT_INITIALIZE, e);
        }
    }

    /**
     * opiQuery: Query device
     * 
     * @return settings and state machine state
     *
     * @since 0.0.1
     */
    public Packet query() {
        if (!this.socket.isConnected()) return Packet.error(DISCONNECTED_FROM_HOST);
        try {
            String q = toJson(Command.QUERY);
            this.send(q);
            Packet rec = this.receive();
            return Packet.checkReturnElements(rec, this.opiMethods, "query");
        } catch (ClassCastException | IllegalArgumentException | IOException e) {
            return Packet.error(COULD_NOT_QUERY, e);
        }
    };

    /**
     * opiSetup: Change device background and overall settings
     * All of the @Parameters here should be fields in the jovp.Setup class.
     * 
     * @param args pairs of argument name and value
     * 
     * @return A JSON object with return messages
     *
     * @since 0.0.1
     */
    @Parameter(name = "eye", className = es.optocom.jovp.definitions.ViewEye.class, desc = "The eye for which to apply the settings.", defaultValue = "BOTH")
    @Parameter(name = "bgLum", className = Double.class, desc = "Background luminance for eye (cd/m^2).", optional = true, min = 0, defaultValue = "128")
    @Parameter(name = "bgCol", className = Double.class, desc = "Background color for eye (rgb).", isList = true, optional = true, min = 0, max = 1, defaultValue = "[1,1,1]")
    @Parameter(name = "bgImageFilename", className = String.class, desc = "If present, display the image in the background for eye", optional = true, defaultValue = "\"\"")
    @Parameter(name = "fixShape", className = es.optocom.jovp.definitions.ModelType.class, desc = "Fixation target type for eye.", optional = true, defaultValue = "MALTESE")
    @Parameter(name = "fixType", className = es.optocom.jovp.definitions.TextureType.class, desc = "Fixation target texture for eye.", optional = true, defaultValue = "FLAT")
    @Parameter(name = "fixImageFilename", className = String.class, desc = "If fixType == IMAGE, the filename on the local filesystem of the machine running JOVP of the image to use", optional = true, defaultValue = "x.jpg")
    @Parameter(name = "fixLum", className = Double.class, desc = "Fixation target luminance for eye.", optional = true, min = 0, defaultValue = "255")
    @Parameter(name = "fixCol", className = Double.class, desc = "Fixation target color for eye.", isList = true, optional = true, min = 0, max = 1, defaultValue = "[0,1,0]")
    @Parameter(name = "fixCx", className = Double.class, desc = "x-coordinate of fixation target (degrees).", optional = true, min = -90, max = 90, defaultValue = "0")
    @Parameter(name = "fixCy", className = Double.class, desc = "y-coordinate of fixation target (degrees).", optional = true, min = -90, max = 90, defaultValue = "0")
    @Parameter(name = "fixSx", className = Double.class, desc = "diameter along major axis of ellipse (degrees). 0 to hide fixation marker.", optional = true, min = 0, defaultValue = "1")
    @Parameter(name = "fixSy", className = Double.class, desc = "diameter along minor axis of ellipse (degrees). If not received, then sy = sx.", optional = true, min = 0, defaultValue = "1")
    @Parameter(name = "fixRotation", className = Double.class, desc = "Angles of rotation of fixation target (degrees). Only useful if sx != sy specified.", optional = true, min = 0, max = 360, defaultValue = "0")
    @Parameter(name = "tracking", className = Integer.class, desc = "Whether to correct stimulus location based on eye position.", optional = true, min = 0, max = 1, defaultValue = "0")
    public Packet setup(HashMap<String, Object> args) {
      if (!this.socket.isConnected()) return Packet.error(DISCONNECTED_FROM_HOST);
      try {
        Packet p = validateArgs(args, this.opiMethods.get("setup").parameters(), "setup");
          if (p.getError()) 
            return(p);
        //this.send(OpiListener.gson.toJson(args));
        this.send(p.getMsg());
        return Packet.checkReturnElements(this.receive(), this.opiMethods, "setup");
      } catch (IOException e) {
        return Packet.error(COULD_NOT_SETUP, e);
      }
    }
   
    /**
     * opiPresent: Present OPI stimulus in perimeter
     * All of the @Parameters should be the fields of the jovp.Present class.  
     *
     * @param args pairs of argument name and value
     * 
     * @return A JSON object with return messages
     *
     * @since 0.0.1
     */
    @Parameter(name = "stim.length", className = Integer.class, desc = "The number of elements in this stimuli.", isList = false, min = 1, defaultValue = "1")
    @Parameter(name = "eye", className = es.optocom.jovp.definitions.ViewEye.class, desc = "The eye for which to apply the settings.", isList = true, defaultValue = "[LEFT]")
    @Parameter(name = "x", className = Double.class, desc = "List of x co-ordinates of stimuli (degrees).", isList = true, min = -90, max = 90, defaultValue = "[0]")
    @Parameter(name = "y", className = Double.class, desc = "List of y co-ordinates of stimuli (degrees).", isList = true, min = -90, max = 90, defaultValue = "[0]")
    @Parameter(name = "sx", className = Double.class, desc = "List of diameters along major axis of ellipse (degrees).", isList = true, min = 0, max = 180, defaultValue = "[1.72]")
    @Parameter(name = "sy", className = Double.class, desc = "List of diameters along minor axis of ellipse (degrees).", isList = true, min = 0, max = 180, defaultValue = "[1.72]")
    @Parameter(name = "t", className = Double.class, desc = "List of stimuli presentation times (ms).", isList = true, min = 0, defaultValue = "[200]")
    @Parameter(name = "w", className = Double.class, desc = "Time to wait for response including presentation time (ms).", isList = false, min = 0, defaultValue = "1500")
    @Parameter(name = "lum", className = Double.class, desc = "List of stimuli luminances (cd/m^2).", isList = true, min = 0, defaultValue = "[300]")
    @Parameter(name = "color1", className = Double.class, desc = "List of stimulus colors for FLAT shapes and patterns.", isListList = true, min = 0, max = 1, defaultValue = "[[0,0,0]]")
    @Parameter(name = "color2", className = Double.class, desc = "List of second colors for non-FLAT shapes", isListList = true, optional = true, min = 0, max = 1, defaultValue = "[[1,1,1]]")
    @Parameter(name = "rotation", className = Double.class, desc = "List of angles of rotation of stimuli (degrees). Only useful if sx != sy specified.", isList = true, optional = true, min = 0, max = 360, defaultValue = "[0]")
    @Parameter(name = "contrast", className = Double.class, desc = "List of stimulus contrasts (from 0 to 1). Only useful if type != FLAT.", isList = true, optional = true, min = 0, max = 1, defaultValue = "[1]")
    @Parameter(name = "phase", className = Double.class, desc = "List of phases (in degrees) for generation of spatial patterns. Only useful if type != FLAT", isList = true, optional = true, min = 0, defaultValue = "[0]")
    @Parameter(name = "frequency", className = Double.class, desc = "List of frequencies (in cycles per degrees) for generation of spatial patterns. Only useful if type != FLAT", isList = true, optional = true, min = 0, max = 300, defaultValue = "[0]")
    @Parameter(name = "envType", className = es.optocom.jovp.definitions.EnvelopeType.class, desc = "List of envelope types to apply to the stims). Only useful if type != FLAT", isList = true, optional = true, defaultValue = "[NONE]")
    @Parameter(name = "envSdx", className = Double.class, desc = "List of envelope sd in x direction in degrees. Only useful if envType != NONE", isList = true, optional = true, defaultValue = "[1.0]")
    @Parameter(name = "envSdy", className = Double.class, desc = "List of envelope sd in y direction in degrees. Only useful if envType != NONE", isList = true, optional = true, defaultValue = "[1.0]")
    @Parameter(name = "envRotation", className = Double.class, desc = "List of envelope rotations in degrees. Only useful if envType != NONE", isList = true, optional = true, defaultValue = "[90]")
    @Parameter(name = "defocus", className = Double.class, desc = "List of defocus values in Diopters for stimulus post-processing.", isList = true, optional = true, min = 0, defaultValue = "[0]")
    @Parameter(name = "texRotation", className = Double.class, desc = "List of angles of rotation of stimuli (degrees). Only useful if type != FLAT", isList = true, optional = true, min = 0, max = 360, defaultValue = "[0]")
    @Parameter(name = "shape", className = es.optocom.jovp.definitions.ModelType.class, desc = "Stimulus shape. Values include CROSS, TRIANGLE, CIRCLE, SQUARE, OPTOTYPE.", isList = true, optional = true, defaultValue = "[CIRCLE]")
    @Parameter(name = "type", className = es.optocom.jovp.definitions.TextureType.class, desc = "Stimulus type. Values include FLAT, SINE, CHECKERBOARD, SQUARESINE, G1, G2, G3, IMAGE", isList = true, optional = true, defaultValue = "[FLAT]")
    @Parameter(name = "imageFilename", className = String.class, desc = "If type == IMAGE, the filename on the local filesystem of the machine running JOVP of the image to use", isList = true, optional = true, defaultValue = "[\"x.jpg\"]")
    @Parameter(name = "fullFoV", className = Double.class, desc = "If !0 fullFoV scales image to full field of view and sx/sy are ignored.", isList = true, optional = true, defaultValue = "[0]")
    @Parameter(name = "optotype", className = es.optocom.jovp.definitions.Optotype.class, desc = "If shape == OPTOTYPE, the letter A to Z to use", isList = true, optional = true, defaultValue = "[E]")
    public Packet present(HashMap<String, Object> args) {
        if (!this.socket.isConnected()) return Packet.error(DISCONNECTED_FROM_HOST);
        try {
            Packet p = validateArgs(args, this.opiMethods.get("present").parameters(), "present");
            if (p.getError()) 
                return(p);
            //this.send(OpiListener.gson.toJson(args));
            this.send(p.getMsg());
            return Packet.checkReturnElements(this.receive(), this.opiMethods, "present");
        } catch (IOException e) {
            return Packet.error(COULD_NOT_PRESENT, e);
        }
    }

    /**
    * opiClose: Send close to Jovp and close my socket to it.
    * 
    * @param args pairs of argument name and value
    *
    * @return A JSON object with return messages
    *
    * @since 0.0.1
    */
    public Packet close() {
        if (!this.socket.isConnected()) return Packet.error(DISCONNECTED_FROM_HOST);
       
        try {
            this.send(toJson(Command.CLOSE));   // this will close the server, so no messages coming back
            this.closeSocket();
        } catch (IOException e) {
            return Packet.error(COULD_NOT_CLOSE, e);
        }
        return new Packet(true, DISCONNECTED_FROM_HOST);
    }

    /** This is what the JOVP Server/Machine is expecting for the Initialize Command */
    private String initConfiguration() {
        return new StringBuilder("{\n  \"command\": " + Command.INITIALIZE + ",\n")
        .append("  \"machine\": " + this.getClass().getSimpleName() + ",\n")
        .append("  \"screen\": " + settings.screen + ",\n")
        .append("  \"physicalSize\": " + Arrays.toString(settings.physicalSize) + ",\n")
        .append("  \"pseudoGray\": " + settings.pseudoGray + ",\n")
        .append("  \"fullScreen\": " + settings.fullScreen + ",\n")
        .append("  \"distance\": " + settings.distance + ",\n")
        .append("  \"viewMode\": " + settings.viewMode + ",\n")
        .append("  \"input\": " + settings.input + ",\n")
        .append("  \"tracking\": " + settings.tracking + ",\n")
        .append("  \"gammaFile\": " + settings.gammaFile)
        .append("\n}").toString();
    }

    /**
    * Get an array of double values from a suitable list
    * 
    * @param list list of objects that can be coerced into doubles
    * 
    * @return an array of doubles
    * 
    * @throws ClassCastException Cast exception
    * 
    * @since 0.0.1
    */
    public static double[] toDoubleArray(Object list) throws ClassCastException {
        return ((ArrayList<?>) list).stream().mapToDouble(Double.class::cast).toArray();
    }
}  

File: rgen/src/main/OPI/R/Jovp.r
# Open Perimetry Interface implementation for Jovp
#
# Copyright [2022] [Andrew Turpin & Ivan Marin-Franch]
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require(rjson)

    # environment for this machine in R
if (exists(".opi_env") && !exists("Jovp", where = .opi_env))
    assign("Jovp", new.env(), envir = .opi_env)

#' Implementation of opiInitialise for the Jovp machine.
#'
#' This is for internal use only. Use [opiInitialise()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#' @param ipMonitor IP Address of the OPI monitor.
#' @param portMonitor TCP port of the OPI monitor.
#' @param ip IP Address of the OPI machine.
#' @param port TCP port of the OPI machine.
#'
#' @return a list contianing:
#'  * res JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - res$error Error code '0' if all good, '1' something wrong.
#'    - res$msg The success or error message.
#'
#' @examples
#' chooseOpi("Jovp")
#' result <- opiInitialise(null)
#'
#' @seealso [opiInitialise()]
#'
opiInitialise_for_Jovp <- function(ipMonitor = NULL, portMonitor = NULL, ip = NULL, port = NULL) {
    assign("socket", open_socket(ipMonitor, portMonitor), .opi_env$Jovp)

    msg <- list(command = "choose", machine = "Jovp")
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$Jovp$socket)
    res <- readLines(.opi_env$Jovp$socket, n = 1)
    print(res)
    res <- rjson::fromJSON(res)

    msg <- list(ipMonitor = ipMonitor, portMonitor = portMonitor, ip = ip, port = port, command = "initialize")
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$Jovp$socket)

    res <- rjson::fromJSON(readLines(.opi_env$Jovp$socket, n = 1))
    return(res)
}

#' Implementation of opiQueryDevice for the Jovp machine.
#'
#' This is for internal use only. Use [opiQueryDevice()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#'
#'
#' @return a list contianing:
#'  * res JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - res$error '0' if success, '1' if error.
#'    - res$msg The error message or a structure with the following data.
#'
#' @examples
#' chooseOpi("Jovp")
#' result <- opiQueryDevice(null)
#'
#' @seealso [opiQueryDevice()]
#'
opiQueryDevice_for_Jovp <- function() {
if(!exists(".opi_env$Jovp") || !exists(".opi_env$Jovp$socket") || is.null(.opi_env$Jovp$socket))
    stop("Cannot call opiQueryDevice without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list()
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$Jovp$socket)

    res <- rjson::fromJSON(readLines(.opi_env$Jovp$socket, n = 1))
    return(res)
}

#' Implementation of opiSetup for the Jovp machine.
#'
#' This is for internal use only. Use [opiSetup()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#' @param eye The eye for which to apply the settings.
#' @param bgLum Background luminance for eye.
#' @param bgCol Background color for eye.
#' @param fixShape Fixation target type for eye.
#' @param fixLum Fixation target luminance for eye.
#' @param fixCol Fixation target color for eye.
#' @param fixCx x-coordinate of fixation target (degrees).
#' @param fixCy y-coordinate of fixation target (degrees).
#' @param fixSx diameter along major axis of ellipse (degrees).
#' @param fixSy diameter along minor axis of ellipse (degrees). If not received,
#'                 then sy = sx.
#' @param fixRotation Angles of rotation of fixation target (degrees). Only
#'                       useful if sx != sy specified.
#' @param tracking Whether to correct stimulus location based on eye position.
#'
#' @return a list contianing:
#'  * res JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - res$error '0' if success, '1' if error.
#'    - res$msg The error message or a structure with the result of QUERY OPI
#'                 command.
#'
#' @examples
#' chooseOpi("Jovp")
#' result <- opiSetup(settings = list(null))
#'
#' @seealso [opiSetup()]
#'
opiSetup_for_Jovp <- function(settings = list(eye = NULL, bgLum = NULL, bgCol = NULL, fixShape = NULL, fixLum = NULL, fixCol = NULL, fixCx = NULL, fixCy = NULL, fixSx = NULL, fixSy = NULL, fixRotation = NULL, tracking = NULL)) {
if(!exists(".opi_env$Jovp") || !exists(".opi_env$Jovp$socket") || is.null(.opi_env$Jovp$socket))
    stop("Cannot call opiSetup without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list(eye = settings$eye, bgLum = settings$bgLum, bgCol = settings$bgCol, fixShape = settings$fixShape, fixLum = settings$fixLum, fixCol = settings$fixCol, fixCx = settings$fixCx, fixCy = settings$fixCy, fixSx = settings$fixSx, fixSy = settings$fixSy, fixRotation = settings$fixRotation, tracking = settings$tracking)
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$Jovp$socket)

    res <- rjson::fromJSON(readLines(.opi_env$Jovp$socket, n = 1))
    return(res)
}

#' Implementation of opiPresent for the Jovp machine.
#'
#' This is for internal use only. Use [opiPresent()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#' @param eye Eye to test.
#' @param shape Stimulus shape.
#' @param type Stimulus type.
#' @param x List of x co-ordinates of stimuli (degrees).
#' @param y List of y co-ordinates of stimuli (degrees).
#' @param sx List of diameters along major axis of ellipse (degrees).
#' @param sy List of diameters along minor axis of ellipse (degrees). If not
#'              received, then sy = sx
#' @param lum List of stimuli luminances (cd/m^2).
#' @param color1 List of stimulus colors 1.
#' @param color2 List of stimulus colors 2. Only useful if stimulus type != FLAT
#' @param rotation List of angles of rotation of stimuli (degrees). Only useful
#'                    if sx != sy specified.
#' @param contrast List of stimulus contrasts (from 0 to 1). Only useful if type
#'                    != FLAT.
#' @param phase List of phases (in degrees) for generation of spatial patterns.
#'                 Only useful if type != FLAT
#' @param frequency List of frequencies (in cycles per degrees) for generation
#'                     of spatial patterns. Only useful if type != FLAT
#' @param defocus List of defocus values in Diopters for stimulus
#'                   post-processing.
#' @param textRotation List of angles of rotation of stimuli (degrees). Only
#'                        useful if type != FLAT
#' @param t List of stimuli presentation times (ms).
#' @param w List of stimuli response windows (ms).
#'
#' @return a list contianing:
#'    - res$error '0' if success, '1' if error.
#'    - res$msg Error message or a structure with the following fields.
#'    - res$msg$seen '1' if seen, '0' if not.
#'    - res$msg$time Response time from stimulus onset if button pressed (ms).
#'    - res$msg$eyex x co-ordinates of pupil at times eyet (degrees).
#'    - res$msg$eyey y co-ordinates of pupil at times eyet (degrees).
#'    - res$msg$eyed Diameter of pupil at times eyet (mm).
#'    - res$msg$eyet Time of (eyex, eyey) pupil from stimulus onset (ms).
#'
#' @examples
#' chooseOpi("Jovp")
#' result <- opiPresent(stim = list(null))
#'
#' @seealso [opiPresent()]
#'
opiPresent_for_Jovp <- function(stim = list(eye = NULL, shape = NULL, type = NULL, x = NULL, y = NULL, sx = NULL, sy = NULL, lum = NULL, color1 = NULL, color2 = NULL, rotation = NULL, contrast = NULL, phase = NULL, frequency = NULL, defocus = NULL, textRotation = NULL, t = NULL, w = NULL)) {
if (!exists(".opi_env") || !exists("Jovp", envir = .opi_env))
    stop("Cannot call opiPresent without an open socket to Monitor. Did you call opiInitialise()?.")
if (!("socket" %in% names(.opi_env$Jovp)) || is.null(.opi_env$Jovp$socket))
    stop("Cannot call opiPresent without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- c(list(command = "present"), lapply(stim, function(p) ifelse(is.null(p), NULL, p)))
    print(msg)
    #msg <- list(command = "present", eye = stim$eye, shape = stim$shape, type = stim$type, x = stim$x, y = stim$y, sx = stim$sx, sy = stim$sy, lum = stim$lum, color1 = stim$color1, color2 = stim$color2, rotation = stim$rotation, contrast = stim$contrast, phase = stim$phase, frequency = stim$frequency, defocus = stim$defocus, textRotation = stim$textRotation, t = stim$t, w = stim$w)
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$Jovp$socket)

    res <- readLines(.opi_env$Jovp$socket, n = 1)
    print(res)
    res <- rjson::fromJSON(res)
    return(res)
}

#' Implementation of opiClose for the Jovp machine.
#'
#' This is for internal use only. Use [opiClose()] with
#' these Arguments and you will get the Value back.
#'
#' @usage NULL
#'
#'
#'
#' @return a list contianing:
#'  * res JSON Object with all of the other fields described in @ReturnMsg
#'           except 'error'.
#'    - res$error '0' if success, '1' if error.
#'    - res$msg The error message or additional results from the CLOSE command
#'
#' @examples
#' chooseOpi("Jovp")
#' result <- opiClose(null)
#'
#' @seealso [opiClose()]
#'
opiClose_for_Jovp <- function() {
if(!exists(".opi_env$Jovp") || !exists(".opi_env$Jovp$socket") || is.null(.opi_env$Jovp$socket))
    stop("Cannot call opiClose without an open socket to Monitor. Did you call opiInitialise()?.")

    msg <- list()
    msg <- rjson::toJSON(msg)
    writeLines(msg, .opi_env$Jovp$socket)

    res <- rjson::fromJSON(readLines(.opi_env$Jovp$socket, n = 1))
    return(res)
}


#' Set background color and luminance in both eyes.
#' Deprecated for OPI >= v3.0.0 and replaced with [opiSetup()].
#' @usage NULL
#' @seealso [opiSetup()]
opiSetBackground_for_Jovp <- function(lum, color, ...) {return("Deprecated")}


